### PR TITLE
Optimise pex distributions some more

### PR DESCRIPTION
**Pull request by peterebden**
Realised that we create a bunch of PexDistributions, each of which has its own regex and zipfile (and we re-iterate the namelist quite a bit).

This cuts it down to one regex, one zipfile, and one pass over the namelist. We still create all the PexDistribution objects but they're lighter-weight when they do their thing.
Only mildly suspicious that not many tests seemed to fail out of the gate...

My test case is reduced from 1m28s to 32s.

### CODE REVIEW

File: .plzconfig.tool_dev [hunk 1]
```diff
@@ -3,7 +3,7 @@
 ;
 ; To use this configuration:
 ; 1) Run `plz build //tools/please_pex`
-; 2) Copy plz-out/bin/tools/please_pex/please_pex to tools/please_pex_dev
+; 2) Run `cp -f plz-out/bin/tools/please_pex/please_pex tools/please_pex_dev`
```
Comment (peterebden) [inline .plzconfig.tool_dev:6]: just a drive-by, it's a lot easier to be able to copy & paste the command here

File: tools/please_pex/pex/pex_main.py [hunk 2]
```diff
@@ -175,18 +168,22 @@ def __init__(self, module_dir=MODULE_DIR):
...
-                        PexDistribution(m.group("name"), sys.argv[0], prefix=module_dir)
-                    )
+                match = r.match(name)
+                if match:
+                    mod = match.group(1)
+                    dist = dists.get(mod)
+                    if not dist:
+                        dist = [PexDistribution(mod, pex_file, zf, prefix=module_dir)]
+                        dists[mod] = dist
+                    dist[0]._files[match.group(2)] = name
```
Comment (toastwaffle) [inline tools/please_pex/pex/pex_main.py:186]: I'm not a fan of directly manipulating a private member of another class. Would it be suboptimal to build the entire set of file mappings before constructing the `PexDistribution`s, or to make the `PexDistribution` construct the file mapping itself?

File: tools/please_pex/pex/pex_main.py [hunk 3]
```diff
@@ -134,24 +134,17 @@ class PexDistribution(Distribution):
...
         self._prefix = prefix
-        self._re = re.compile(r"{path}(?:-.*)?\.(?:dist|egg)-info/(.*)".format(
-            path=os.path.join(self._prefix, self._name) if self._prefix else self._name,
-        ))
-
-    def _match_file(self, name, filename):
-        match = self._re.match(name)
-        if match and match.group(1) == filename:
-            return name
+        self._files = {}
```
Comment (toastwaffle) [inline tools/please_pex/pex/pex_main.py:142]: Could probably do with a comment somewhere to explain what this is

File: tools/please_pex/pex/pex_main.py [hunk 4]
```diff
@@ -175,18 +168,22 @@ def __init__(self, module_dir=MODULE_DIR):
 
     def _find_all_distributions(self, module_dir):
         dists = {}
-        if zipfile.is_zipfile(sys.argv[0]):
-            zf = ZipFileWithPermissions(sys.argv[0])
+        pex_file = sys.argv[0]
+        if zipfile.is_zipfile(pex_file):
```
Comment (toastwaffle) [inline tools/please_pex/pex/pex_main.py:172]: Sidebar since it was already here: is there any world in which the input is not a zipfile? Should we either log/raise in an else clause, or just remove the `if` and let an exception happen?

**Review by toastwaffle — APPROVED** [review]
I don't feel particularly strongly about these comments, so approving anyway

File: tools/please_pex/pex/pex_main.py (hunk 4 above)
Comment (chrisnovakovic) [reply tools/please_pex/pex/pex_main.py:172]: I _think_ (although I'm not sure, because this isn't really documented very well upstream) that some `MetaPathFinder` abstract methods are supposed to return `None` if the object can't successfully answer a given query. We've had bad experiences recently when raising exceptions in this code. If we put a log line in here, I suspect it'll be spamming stderr constantly.

Comment (peterebden) [reply tools/please_pex/pex/pex_main.py:172]: Yes, that matches my understanding; MetaPathFinders don't have to handle a particular request. I agree that logging would be too noisy.

I think it might turn out not to be a zipfile if the pex was marked zip-unsafe; or maybe there is some other edge case (e.g. what if people get adventurous with `multiprocessing`, I've seen some weird outcomes there)

**Review by chrisnovakovic — APPROVED** [review]

File: tools/please_pex/pex/pex_main.py (hunk 2 above)
Comment (peterebden) [reply tools/please_pex/pex/pex_main.py]: Yeah fair. It is a bit cheeky, although it's also a 'local' class so arguably they are allowed to know about each other a bit.
Anyway, have built up the files then gone back to do the distributions in one pass; I think it is a bit nicer like this.
